diff -up sysstat-9.0.4/rd_stats.c.res-leak sysstat-9.0.4/rd_stats.c
--- sysstat-9.0.4/rd_stats.c.res-leak	2012-02-16 15:52:01.000000000 +0100
+++ sysstat-9.0.4/rd_stats.c	2012-02-16 15:58:06.445877640 +0100
@@ -1897,8 +1897,10 @@ void read_uptime(unsigned long long *upt
 	if ((fp = fopen(UPTIME, "r")) == NULL)
 		return;
 
-	if (fgets(line, 128, fp) == NULL)
+	if (fgets(line, 128, fp) == NULL) {
+		fclose(fp);
 		return;
+	}
 
 	sscanf(line, "%lu.%lu", &up_sec, &up_cent);
 	*uptime = (unsigned long long) up_sec * HZ + (unsigned long long) up_cent * HZ / 100;
