diff -upr sysstat-9.0.4.old/common.c sysstat-9.0.4/common.c
--- sysstat-9.0.4.old/common.c	2012-02-08 13:34:56.412784672 +0100
+++ sysstat-9.0.4/common.c	2012-02-08 13:34:34.173139712 +0100
@@ -378,10 +378,10 @@ int get_win_height(void)
 
 /*
  ***************************************************************************
- * Remove /dev from path name.
+ * Canonicalize and remove /dev from path name.
  *
  * IN:
- * @name	Device name (may begins with "/dev/")
+ * @name       Device name (may begins with "/dev/" or can be a symlink).
  *
  * RETURNS:
  * Device basename.
@@ -389,10 +389,27 @@ int get_win_height(void)
  */
 char *device_name(char *name)
 {
-	if (!strncmp(name, "/dev/", 5))
-		return name + 5;
+	static char out[MAX_FILE_LEN];
+	char *resolved_name;
+	int i = 0;
+
+	/* realpath() creates new string, so we need to free it later */
+	resolved_name = realpath(name, NULL);
+
+	/* If path doesn't exist, just return input. */
+	if (!resolved_name) {
+		return name;
+	}
+
+	if (!strncmp(resolved_name, "/dev/", 5)) {
+		i = 5;
+	}
+	strncpy(out, resolved_name + i, MAX_FILE_LEN);
+	out[MAX_FILE_LEN - 1] = '\0';
+
+	free(resolved_name);
 
-	return name;
+	return out; 
 }
 
 /*
