diff -ur sysstat-9.0.4.orig/sa1.in sysstat-9.0.4/sa1.in
--- sysstat-9.0.4.orig/sa1.in	2013-08-13 16:57:56.627270000 +0200
+++ sysstat-9.0.4/sa1.in	2013-08-13 17:11:41.202088133 +0200
@@ -8,6 +8,7 @@
 HISTORY=0
 SADC_OPTIONS="-S DISK"
 SYSCONFIG_DIR=@SYSCONFIG_DIR@
+umask 0022
 [ -r ${SYSCONFIG_DIR}/sysstat ] && . ${SYSCONFIG_DIR}/sysstat
 if [ ${HISTORY} -gt 28 ]
 then
@@ -30,7 +31,6 @@
 	rm -f ${CURRENTFILE}
 	ln -s ${CURRENTDIR}/${CURRENTFILE} ${CURRENTFILE}
 fi
-umask 0022
 ENDIR=@SA_LIB_DIR@
 cd ${ENDIR}
 [ "$1" = "--boot" ] && shift && BOOT=y || BOOT=n
